src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,354 @@
+{-# LANGUAGE BangPatterns    #-}
+{-# LANGUAGE LambdaCase      #-}
+{-# LANGUAGE NamedFieldPuns  #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | Datastructures and algorithms for k-way merges.
+--
+-- === TODO
+--
+-- This is temporary module header documentation. The module will be
+-- fleshed out more as we implement bits of it.
+--
+-- Related work packages: 5, 6
+--
+-- This module includes in-memory parts for, amongst others,
+--
+-- * k-way merge data structures (heap or tournament tree)
+--
+-- * k-way merge algorithms
+--
+-- The above list is a sketch. Functionality may move around, and the list is
+-- not exhaustive.
+--
+module Database.LSMTree.Internal.Merge (
+    Merge (..)
+  , new
+  , MergeValues
+  , StepResult (..)
+  , steps
+  , close
+  ) where
+
+import           Control.Exception (assert)
+import           Control.Monad (zipWithM)
+import           Control.Monad.Primitive (RealWorld)
+import           Data.Bifunctor (first)
+import           Data.Function (on)
+import           Data.IORef
+import           Data.Maybe (catMaybes)
+import           Data.Traversable (for)
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.BitMath
+import           Database.LSMTree.Internal.Entry
+import           Database.LSMTree.Internal.PageAcc (entryWouldFitInPage)
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.RawOverflowPage (RawOverflowPage,
+                     rawOverflowPageRawBytes)
+import           Database.LSMTree.Internal.RawPage
+import           Database.LSMTree.Internal.Run (Run)
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Run.Mutable (MRun)
+import qualified Database.LSMTree.Internal.Run.Mutable as MRun
+import           Database.LSMTree.Internal.Run.Reader (Reader)
+import qualified Database.LSMTree.Internal.Run.Reader as Reader
+import           Database.LSMTree.Internal.Serialise
+import qualified KMerge.Heap as Heap
+import qualified System.FS.API as FS
+import           System.FS.API (HasBufFS, HasFS)
+
+-- | An in-progress incremental k-way merge of 'Run's.
+--
+-- Each step reads exactly one entry and potentially writes one entry.
+--
+-- TODO: Try the other approach:
+-- Always produce one entry (which could use multiple inputs of the same key).
```

Comment:
Yes, although we could track how many inputs we read and still stop at pretty much the right moment. If having read a hand full of entries too many (bounded by the number of input runs) is a problem, we can even track that to make sure we read a little less next time.
It might even be that finishing the merge early is not a bad thing, since it only really happens when there are many redundant entries, so we get a lot of compaction, which is nice to benefit from sooner rather than later. Of course the write overhead becomes a little bit uneven.

Reply:
On the other hand I think the "always write 1 entry" approach is going to be faster and make some other optimisations easier to do.

Reply:
I want to try it once there is a benchmark.